### PR TITLE
Resolve GH180, GH176 and GH166

### LINKED ISSUE #166: v2.2.4 spams to log with NSLog on iOs
After update to the latest version I see huge amount of annoying logs produced by following line:

https://github.com/rollbar/rollbar-apple/blob/9789a2d161f8d608bd19bdf7db4119bdf7cb2bfa/RollbarNotifier/Sources/RollbarNotifier/RollbarLogger.m#L1034

Is it intended?

### LINKED ISSUE #176: Rollbar crashes when device runs out of memory
**Describe the bug**
It looks like `RollbarLogger.m` contains code that is using deprecated [writeData:](https://developer.apple.com/documentation/foundation/nsfilehandle/1410936-writedata) method that does not handle errors gracefully. Therefore when the device runs out of space, the method throws an exception. 

Here is the crash report from the Crashlytics. 
```
*** -[NSConcreteFileHandle writeData:]: No space left on device

Fatal Exception: NSFileHandleOperationException
0  CoreFoundation                 0x91d3c __exceptionPreprocess
1  libobjc.A.dylib                0x146a8 objc_exception_throw
2  Foundation                     0x125264 -[NSConcreteFileHandle offsetInFile]
3  Foundation                     0x5dff0 -[NSConcreteFileHandle writeData:]
4  RollbarNotifier                0xf164 -[RollbarLogger queuePayload_OnlyCallOnThread:] + 736 (RollbarLogger.m:736)
5  Foundation                     0x3a868 __NSThreadPerformPerform
6  CoreFoundation                 0xb22b0 __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__
7  CoreFoundation                 0xc22d8 __CFRunLoopDoSource0
8  CoreFoundation                 0x5678 __CFRunLoopDoSources0
9  CoreFoundation                 0xaaa8 __CFRunLoopRun
10 CoreFoundation                 0x1de1c CFRunLoopRunSpecific
11 Foundation                     0x173dc -[NSRunLoop(NSRunLoop) runMode:beforeDate:]
12 RollbarNotifier                0x1925c -[RollbarThread run] + 71 (RollbarThread.m:71)
13 Foundation                     0x638bc __NSThread__start__
14 libsystem_pthread.dylib        0x63a4 _pthread_start
15 libsystem_pthread.dylib        0x49fc thread_start
```

**To Reproduce**
I didn't try to reproduce the error, but I think that if there's a way to fill up the simulator space, it is possible to trigger the error. 

**Expected behavior**
App should not crash

**Rollbar-Apple SDK version: v2.0.3**

**Calling the SDK from (x-mark all suitable):**
- [ ] Objective-C
- [x] Swift
- [ ] Other: 

**Smartphone (please complete the following information):**
 - Device: iPhone X
 - OS iOS 15.3.1

### LINKED ISSUE #180: SonarCloud: address current reliability issues

### PR DESCRIPTION
**Pull request by akornich**
## Description of the change

> resolve #180
> resolve #176
>resolve #166 

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Maintenance
- [ ] New release

## Related issues

> Shortcut stories and GitHub issues (delete irrelevant)

- Fix #180 
- Fix #176
- Fix #166 

### Development

- [x] Lint rules pass locally
- [] The code changed/added as part of this pull request has been covered with tests
- [x] All tests related to the changed code pass in development

### Code review

- [ ] This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers assigned
- [ ] Issue from task tracker has a link to this pull request
- [ ] Changes have been reviewed by at least one other engineer

### CODE REVIEW

File: RollbarCommon/Sources/RollbarCommon/DTOs/RollbarDTO+Protected.m [hunk 1]
```diff
@@ -116,19 +116,29 @@ - (NSNumber *)safelyGetNumberByKey:(NSString *)key {
 
 #pragma mark - data setters by key
 
-- (void)setDictionary:(NSDictionary *)data forKey:(NSString *)key {
-    [self setData:data.mutableCopy byKey:key];
+- (void)setDictionary:(nullable NSDictionary *)data forKey:(NSString *)key {
+    
+    [self setData:!data ? data.mutableCopy : data
```
[the remark this answers is not in the document]
Comment (akornich) [reply RollbarCommon/Sources/RollbarCommon/DTOs/RollbarDTO+Protected.m:121]: @matux , good catch. However, it looks like NULL is accepted as nil once inside this method. I also added a unit test to make sure it is always true.

[the remark this answers is not in the document]
Comment (matux) [reply RollbarCommon/Sources/RollbarCommon/DTOs/RollbarDTO+Protected.m:121]: Yeah, `NULL` and `nil` are the same thing in Obj-C. It's just you're calling `mutableCopy` on a `nil`, which will return `nil`, what for?

Why not just pass `data`?

[the remark this answers is not in the document]
Comment (akornich) [reply RollbarCommon/Sources/RollbarCommon/DTOs/RollbarDTO+Protected.m]: Oh, I see. I did not even pay attention to !data the first time I looked at your comment. That is a typo. In fact, there were more. Fixed all now. Good catch! Thank you!

**Review by matux — APPROVED** [review]
Approved

**Comment by akornich** [comment]
@matux , thank you!

**Pull request merged**